lib/components/OverdriveProvider/OverdriveProvider.tsx

```diff
@@ -1,43 +1,134 @@
 }
 
-export const OverdriveProvider = ({
-	noBodyLevelTheming = false,
+export interface ProviderContext
+	extends Pick<ProviderProps, 'colorOverrides' | 'portalMountPoint'> {
+	themeClass: (typeof baseTheme)['themeRef'];
+	themeName: string;
+	overrideStyles: ReturnType<typeof assignInlineVars>;
+	vars: ThemeContract;
+}
+
+const OverdriveContext = createContext<ProviderContext | null>(null);
+const RuntimeTokensContext = createContext<RuntimeTokens | null>(null);
+const msgInvariantError = "You haven't used an `<OverdriveProvider>`.";
+
+export const useTheme = () => {
+	const context = useContext(OverdriveContext);
+	invariant(context !== null, msgInvariantError);
+	return context;
+};
+
+export const useRuntimeTokens = (): RuntimeTokens => {
+	const tokens = useContext(RuntimeTokensContext);
+	invariant(tokens !== null, msgInvariantError);
+	return tokens;
+};
+
+const isSafeElement = (element: HTMLElement | null): boolean => {
+	return (
+		element instanceof HTMLElement &&
+		!element.hasAttribute('onclick') &&
+		!element.hasAttribute('onerror')
+	);
+};
+
+export const Provider = ({
+	breakpoints,
 	children,
-	...props
-}: OverdriveProviderProps) => {
+	colorOverrides,
+	noBodyLevelTheming = false,
+	portalMountPoint,
+	theme = baseTheme,
+}: ProviderProps) => {
+	const runtimeTokens = useMemo(
+		() => makeRuntimeTokens(breakpoints),
+		[breakpoints],
+	);
+	const styles = useColorOverrides(colorOverrides, String(theme.vars.mode));
+	const themeValues = useMemo(
+		() => ({
+			themeClass: theme.themeRef,
+			themeName: theme.name,
+			colorOverrides,
+			overrideStyles: styles,
+			portalMountPoint,
+			vars: theme.vars,
+		}),
+		[colorOverrides, portalMountPoint, styles, theme],
+	);
+
+	// Body Level Theming
 	useEffect(() => {
 		if (
 			!isBrowser ||
 			noBodyLevelTheming ||
-			document.body.classList.contains(props.themeClass)
+			document.body.classList.contains(theme.themeRef)
 		)
 			return;
 
-		// Body has theme class applied to it, so we use css vars to apply body styles
-		document.body.classList.add(props.themeClass);
-
-		// Cleanup function to remove the class when component unmounts
+		document.body.classList.add(theme.themeRef);
 		return () => {
-			document.body.classList.remove(props.themeClass);
+			document.body.classList.remove(theme.themeRef);
 		};
-	}, [noBodyLevelTheming, props.themeClass]);
+	}, [noBodyLevelTheming, theme]);
+
+	// Check portal mount point ref
+	useEffect(() => {
+		if (
+			portalMountPoint?.current &&
+			!isSafeElement(portalMountPoint.current)
+		) {
+			console.error('Unsafe portal mount point detected');
+			portalMountPoint.current = null;
+		}
+	}, [portalMountPoint]);
 
-	return <ThemeProvider {...props}>{children}</ThemeProvider>;
+	return (
+		<OverdriveContext.Provider value={themeValues}>
+			<RuntimeTokensContext.Provider value={runtimeTokens}>
+				<div
+					className={theme.themeRef}
+					style={styles}
+					data-od-component-provider=""
+				>
```

Comment:
You may consider putting a `display: contents` on this div, so the using of the provider does not cause unintended, layout reflows Providers, at least in React, typically carry with them the idea that they do not "render anything". 

Though fully accept you may want a discrete block-level div for this boundary.